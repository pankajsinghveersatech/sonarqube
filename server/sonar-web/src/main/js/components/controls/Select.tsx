/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import ReactSelectClass, {
  ReactAsyncSelectProps,
  ReactCreatableSelectProps,
  ReactSelectProps
} from 'react-select';
import { lazyLoadComponent } from '../lazyLoadComponent';
import { ClearButton } from './buttons';
import './Select.css';

const ReactSelectLib = import('react-select');
const ReactSelect = lazyLoadComponent(() => ReactSelectLib);
const ReactCreatable = lazyLoadComponent(() =>
  ReactSelectLib.then(lib => ({ default: lib.Creatable }))
);
const ReactAsync = lazyLoadComponent(() => ReactSelectLib.then(lib => ({ default: lib.Async })));

function renderInput() {
  return <ClearButton className="button-tiny spacer-left text-middle" iconProps={{ size: 12 }} />;
}

export interface WithInnerRef {
  innerRef?: React.Ref<ReactSelectClass<unknown>>;
}

export default function Select({ innerRef, ...props }: WithInnerRef & ReactSelectProps) {
  // hide the "x" icon when select is empty
  const clearable = props.clearable ? Boolean(props.value) : false;
  return (
    <ReactSelect {...props} clearable={clearable} clearRenderer={renderInput} ref={innerRef} />
  );
}

export function Creatable(props: ReactCreatableSelectProps) {
  return <ReactCreatable {...props} clearRenderer={renderInput} />;
}

export function AsyncSelect(props: ReactAsyncSelectProps & WithInnerRef) {
  return <ReactAsync {...props} />;
}
