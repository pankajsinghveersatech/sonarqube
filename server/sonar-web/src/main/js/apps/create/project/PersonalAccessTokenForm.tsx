/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import {
  checkPersonalAccessTokenIsValid,
  setAlmPersonalAccessToken
} from '../../../api/alm-integrations';
import { SubmitButton } from '../../../components/controls/buttons';
import ValidationInput from '../../../components/controls/ValidationInput';
import { Alert } from '../../../components/ui/Alert';
import DeferredSpinner from '../../../components/ui/DeferredSpinner';
import { translate } from '../../../helpers/l10n';
import { getBaseUrl } from '../../../helpers/system';
import { AlmKeys, AlmSettingsInstance } from '../../../types/alm-settings';

interface Props {
  almSetting: AlmSettingsInstance;
  resetPat: boolean;
  onPersonalAccessTokenCreated: () => void;
}

interface State {
  validationFailed: boolean;
  validationErrorMessage?: string;
  touched: boolean;
  password: string;
  username?: string;
  submitting: boolean;
  checkingPat: boolean;
}

function getPatUrl(alm: AlmKeys, url = '') {
  if (alm === AlmKeys.BitbucketServer) {
    return `${url.replace(/\/$/, '')}/plugins/servlet/access-tokens/add`;
  } else if (alm === AlmKeys.BitbucketCloud) {
    return 'https://bitbucket.org/account/settings/app-passwords/new';
  } else if (alm === AlmKeys.GitLab) {
    return 'https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html';
  }

  return '';
}

export default class PersonalAccessTokenForm extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);

    this.state = {
      checkingPat: false,
      touched: false,
      password: '',
      submitting: false,
      validationFailed: false
    };
  }

  async componentDidMount() {
    const {
      almSetting: { key },
      resetPat
    } = this.props;
    this.mounted = true;

    // We don't need to check PAT if we want to reset
    if (!resetPat) {
      this.setState({ checkingPat: true });
      const { patIsValid, error } = await checkPersonalAccessTokenIsValid(key)
        .then(({ status, error }) => ({ patIsValid: status, error }))
        .catch(() => ({ patIsValid: status, error: translate('default_error_message') }));
      if (patIsValid) {
        this.props.onPersonalAccessTokenCreated();
      }
      if (this.mounted) {
        // This is the initial message when no token was provided
        if (error === `personal access token for '${key}' is missing`) {
          this.setState({
            checkingPat: false
          });
        } else {
          this.setState({
            checkingPat: false,
            validationFailed: true,
            validationErrorMessage: error
          });
        }
      }
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleUsernameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({
      touched: true,
      username: event.target.value
    });
  };

  handlePasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({
      touched: true,
      password: event.target.value
    });
  };

  handleSubmit = async (e: React.SyntheticEvent<HTMLFormElement>) => {
    const { password, username } = this.state;
    const {
      almSetting: { key }
    } = this.props;

    e.preventDefault();
    if (password) {
      this.setState({ submitting: true });

      await setAlmPersonalAccessToken(key, password, username).catch(() => {
        /* Set will not check pat validity. We need to check again so we will catch issue after */
      });

      const { status, error } = await checkPersonalAccessTokenIsValid(key)
        .then(({ status, error }) => ({ status, error }))
        .catch(() => ({ status: false, error: translate('default_error_message') }));

      if (this.mounted && status) {
        // Let's reset status,
        this.setState({
          checkingPat: false,
          touched: false,
          password: '',
          submitting: false,
          username: '',
          validationFailed: false
        });
        this.props.onPersonalAccessTokenCreated();
      } else if (this.mounted) {
        this.setState({
          submitting: false,
          touched: false,
          validationFailed: true,
          validationErrorMessage: error
        });
      }
    }
  };

  render() {
    const {
      almSetting: { alm, url }
    } = this.props;
    const {
      checkingPat,
      submitting,
      touched,
      password,
      username,
      validationFailed,
      validationErrorMessage
    } = this.state;

    if (checkingPat) {
      return <DeferredSpinner className="spacer-left" loading={true} />;
    }

    const suffixTranslationKey = alm === AlmKeys.BitbucketCloud ? '.bitbucketcloud' : '';

    const isInvalid = validationFailed && !touched;
    const canSubmit = Boolean(password) && (alm !== AlmKeys.BitbucketCloud || Boolean(username));
    const submitButtonDiabled = isInvalid || submitting || !canSubmit;

    const errorMessage =
      validationErrorMessage ?? translate('onboarding.create_project.pat_incorrect', alm);

    return (
      <div className="display-flex-start">
        <form className="width-50" onSubmit={this.handleSubmit}>
          <h2 className="big">{translate('onboarding.create_project.pat_form.title', alm)}</h2>
          <p className="big-spacer-top big-spacer-bottom">
            {translate('onboarding.create_project.pat_form.help', alm)}
          </p>

          {alm === AlmKeys.BitbucketCloud && (
            <ValidationInput
              error={undefined}
              id="enter_username_validation"
              isInvalid={false}
              isValid={false}
              label={translate('onboarding.create_project.enter_username')}
              required={true}>
              <input
                autoFocus={true}
                className={classNames('input-super-large', {
                  'is-invalid': isInvalid
                })}
                id="username"
                minLength={1}
                name="username"
                value={username}
                onChange={this.handleUsernameChange}
                type="text"
              />
            </ValidationInput>
          )}

          <ValidationInput
            error={errorMessage}
            id="personal_access_token_validation"
            isInvalid={false}
            isValid={false}
            label={translate(`onboarding.create_project.enter_pat${suffixTranslationKey}`)}
            required={true}>
            <input
              autoFocus={alm !== AlmKeys.BitbucketCloud}
              className={classNames('input-super-large', {
                'is-invalid': isInvalid
              })}
              id="personal_access_token"
              minLength={1}
              value={password}
              onChange={this.handlePasswordChange}
              type="text"
            />
          </ValidationInput>

          <ValidationInput
            error={errorMessage}
            id="personal_access_token_submit"
            isInvalid={isInvalid}
            isValid={false}
            label={null}>
            <SubmitButton disabled={submitButtonDiabled}>{translate('save')}</SubmitButton>
            <DeferredSpinner className="spacer-left" loading={submitting} />
          </ValidationInput>
        </form>

        <Alert className="big-spacer-left width-50" display="block" variant="info">
          {alm === AlmKeys.BitbucketCloud && (
            <>
              <h3>
                {translate(
                  'onboarding.create_project.pat_help.instructions_username.bitbucketcloud.title'
                )}
              </h3>
              <p className="big-spacer-top big-spacer-bottom">
                {translate(
                  'onboarding.create_project.pat_help.instructions_username.bitbucketcloud'
                )}
              </p>

              <div className="text-middle big-spacer-bottom">
                <img
                  alt="" // Should be ignored by screen readers
                  className="spacer-right"
                  height="16"
                  src={`${getBaseUrl()}/images/alm/${AlmKeys.BitbucketServer}.svg`}
                />
                <a
                  href="https://bitbucket.org/account/settings/"
                  rel="noopener noreferrer"
                  target="_blank">
                  {translate(
                    'onboarding.create_project.pat_help.instructions_username.bitbucketcloud.link'
                  )}
                </a>
              </div>
            </>
          )}

          <h3>{translate(`onboarding.create_project.pat_help${suffixTranslationKey}.title`)}</h3>

          <p className="big-spacer-top big-spacer-bottom">
            <FormattedMessage
              id="onboarding.create_project.pat_help.instructions"
              defaultMessage={translate(
                `onboarding.create_project.pat_help${suffixTranslationKey}.instructions`
              )}
              values={{ alm: translate('onboarding.alm', alm) }}
            />
          </p>

          {(url || alm === AlmKeys.BitbucketCloud) && (
            <div className="text-middle">
              <img
                alt="" // Should be ignored by screen readers
                className="spacer-right"
                height="16"
                src={`${getBaseUrl()}/images/alm/${
                  alm === AlmKeys.BitbucketCloud ? AlmKeys.BitbucketServer : alm
                }.svg`}
              />
              <a href={getPatUrl(alm, url)} rel="noopener noreferrer" target="_blank">
                {translate(`onboarding.create_project.pat_help${suffixTranslationKey}.link`)}
              </a>
            </div>
          )}

          <p className="big-spacer-top big-spacer-bottom">
            {translate('onboarding.create_project.pat_help.instructions2', alm)}
          </p>

          <ul>
            {alm === AlmKeys.BitbucketServer && (
              <li>
                <FormattedMessage
                  defaultMessage={translate(
                    'onboarding.create_project.pat_help.bbs_permission_projects'
                  )}
                  id="onboarding.create_project.pat_help.bbs_permission_projects"
                  values={{
                    perm: (
                      <strong>
                        {translate('onboarding.create_project.pat_help.read_permission')}
                      </strong>
                    )
                  }}
                />
              </li>
            )}
            {(alm === AlmKeys.BitbucketServer || alm === AlmKeys.BitbucketCloud) && (
              <li>
                <FormattedMessage
                  defaultMessage={translate(
                    'onboarding.create_project.pat_help.bbs_permission_repos'
                  )}
                  id="onboarding.create_project.pat_help.bbs_permission_repos"
                  values={{
                    perm: (
                      <strong>
                        {translate('onboarding.create_project.pat_help.read_permission')}
                      </strong>
                    )
                  }}
                />
              </li>
            )}

            {alm === AlmKeys.GitLab && (
              <li className="spacer-bottom">
                <strong>
                  {translate('onboarding.create_project.pat_help.gitlab.read_api_permission')}
                </strong>
              </li>
            )}
          </ul>
        </Alert>
      </div>
    );
  }
}
