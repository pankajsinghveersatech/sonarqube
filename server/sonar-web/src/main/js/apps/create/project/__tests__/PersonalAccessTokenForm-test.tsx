/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import {
  checkPersonalAccessTokenIsValid,
  setAlmPersonalAccessToken
} from '../../../../api/alm-integrations';
import { SubmitButton } from '../../../../components/controls/buttons';
import {
  mockAlmSettingsInstance,
  mockBitbucketCloudAlmSettingsInstance
} from '../../../../helpers/mocks/alm-settings';
import { change, submit, waitAndUpdate } from '../../../../helpers/testUtils';
import { AlmKeys } from '../../../../types/alm-settings';
import PersonalAccessTokenForm from '../PersonalAccessTokenForm';

jest.mock('../../../../api/alm-integrations', () => ({
  checkPersonalAccessTokenIsValid: jest.fn().mockResolvedValue({ status: true }),
  setAlmPersonalAccessToken: jest.fn().mockResolvedValue({})
}));

it('should render correctly', async () => {
  expect(shallowRender()).toMatchSnapshot('no token needed');

  (checkPersonalAccessTokenIsValid as jest.Mock).mockResolvedValueOnce({ status: false });
  let wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('bitbucket');

  (checkPersonalAccessTokenIsValid as jest.Mock).mockResolvedValueOnce({ status: false });
  wrapper = shallowRender({ almSetting: mockBitbucketCloudAlmSettingsInstance() });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('bitbucket cloud');

  (checkPersonalAccessTokenIsValid as jest.Mock).mockResolvedValueOnce({ status: false });
  wrapper = shallowRender({
    almSetting: mockAlmSettingsInstance({ alm: AlmKeys.GitLab, url: 'https://gitlab.com/api/v4' })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('gitlab');

  (checkPersonalAccessTokenIsValid as jest.Mock).mockResolvedValueOnce({ status: false });
  wrapper = shallowRender({
    almSetting: mockAlmSettingsInstance({
      alm: AlmKeys.GitLab,
      url: 'https://gitlabapi.unexpectedurl.org'
    })
  });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('gitlab with non-standard api path');
});

it('should correctly handle form interactions', async () => {
  const onPersonalAccessTokenCreated = jest.fn();
  const wrapper = shallowRender({ onPersonalAccessTokenCreated });

  await waitAndUpdate(wrapper);
  // Submit button disabled by default.
  expect(wrapper.find(SubmitButton).prop('disabled')).toBe(true);

  // Submit button enabled if there's a value.
  change(wrapper.find('input'), 'token');
  expect(wrapper.find(SubmitButton).prop('disabled')).toBe(false);

  // Expect correct calls to be made when submitting.
  submit(wrapper.find('form'));
  expect(onPersonalAccessTokenCreated).toBeCalled();
  expect(setAlmPersonalAccessToken).toBeCalledWith('key', 'token', undefined);
});

it('should correctly handle form for bitbucket interactions', async () => {
  const onPersonalAccessTokenCreated = jest.fn();
  const wrapper = shallowRender({
    almSetting: mockBitbucketCloudAlmSettingsInstance(),
    onPersonalAccessTokenCreated
  });

  await waitAndUpdate(wrapper);
  // Submit button disabled by default.
  expect(wrapper.find(SubmitButton).prop('disabled')).toBe(true);

  change(wrapper.find('#personal_access_token'), 'token');
  expect(wrapper.find(SubmitButton).prop('disabled')).toBe(true);

  // Submit button enabled if there's a value.
  change(wrapper.find('#username'), 'username');
  expect(wrapper.find(SubmitButton).prop('disabled')).toBe(false);

  // Expect correct calls to be made when submitting.
  submit(wrapper.find('form'));
  expect(onPersonalAccessTokenCreated).toBeCalled();
  expect(setAlmPersonalAccessToken).toBeCalledWith('key', 'token', 'username');
});

it('should show error when issue', async () => {
  (checkPersonalAccessTokenIsValid as jest.Mock).mockRejectedValueOnce({});
  const wrapper = shallowRender({
    almSetting: mockBitbucketCloudAlmSettingsInstance()
  });

  await waitAndUpdate(wrapper);

  (checkPersonalAccessTokenIsValid as jest.Mock).mockRejectedValueOnce({});

  change(wrapper.find('#personal_access_token'), 'token');
  change(wrapper.find('#username'), 'username');

  // Expect correct calls to be made when submitting.
  submit(wrapper.find('form'));
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('issue submitting token');
});

function shallowRender(props: Partial<PersonalAccessTokenForm['props']> = {}) {
  return shallow<PersonalAccessTokenForm>(
    <PersonalAccessTokenForm
      almSetting={mockAlmSettingsInstance({
        alm: AlmKeys.BitbucketServer,
        url: 'http://www.example.com'
      })}
      onPersonalAccessTokenCreated={jest.fn()}
      resetPat={false}
      {...props}
    />
  );
}
