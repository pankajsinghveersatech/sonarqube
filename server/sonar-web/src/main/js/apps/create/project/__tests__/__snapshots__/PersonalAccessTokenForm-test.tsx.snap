// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: bitbucket 1`] = `
<div
  className="display-flex-start"
>
  <form
    className="width-50"
    onSubmit={[Function]}
  >
    <h2
      className="big"
    >
      onboarding.create_project.pat_form.title.bitbucket
    </h2>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_form.help.bitbucket
    </p>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.bitbucket"
      id="personal_access_token_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_pat"
      required={true}
    >
      <input
        autoFocus={true}
        className="input-super-large is-invalid"
        id="personal_access_token"
        minLength={1}
        onChange={[Function]}
        type="text"
        value=""
      />
    </ValidationInput>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.bitbucket"
      id="personal_access_token_submit"
      isInvalid={true}
      isValid={false}
      label={null}
    >
      <SubmitButton
        disabled={true}
      >
        save
      </SubmitButton>
      <DeferredSpinner
        className="spacer-left"
        loading={false}
      />
    </ValidationInput>
  </form>
  <Alert
    className="big-spacer-left width-50"
    display="block"
    variant="info"
  >
    <h3>
      onboarding.create_project.pat_help.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      <FormattedMessage
        defaultMessage="onboarding.create_project.pat_help.instructions"
        id="onboarding.create_project.pat_help.instructions"
        values={
          Object {
            "alm": "onboarding.alm.bitbucket",
          }
        }
      />
    </p>
    <div
      className="text-middle"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/bitbucket.svg"
      />
      <a
        href="http://www.example.com/plugins/servlet/access-tokens/add"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.link
      </a>
    </div>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions2.bitbucket
    </p>
    <ul>
      <li>
        <FormattedMessage
          defaultMessage="onboarding.create_project.pat_help.bbs_permission_projects"
          id="onboarding.create_project.pat_help.bbs_permission_projects"
          values={
            Object {
              "perm": <strong>
                onboarding.create_project.pat_help.read_permission
              </strong>,
            }
          }
        />
      </li>
      <li>
        <FormattedMessage
          defaultMessage="onboarding.create_project.pat_help.bbs_permission_repos"
          id="onboarding.create_project.pat_help.bbs_permission_repos"
          values={
            Object {
              "perm": <strong>
                onboarding.create_project.pat_help.read_permission
              </strong>,
            }
          }
        />
      </li>
    </ul>
  </Alert>
</div>
`;

exports[`should render correctly: bitbucket cloud 1`] = `
<div
  className="display-flex-start"
>
  <form
    className="width-50"
    onSubmit={[Function]}
  >
    <h2
      className="big"
    >
      onboarding.create_project.pat_form.title.bitbucketcloud
    </h2>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_form.help.bitbucketcloud
    </p>
    <ValidationInput
      id="enter_username_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_username"
      required={true}
    >
      <input
        autoFocus={true}
        className="input-super-large is-invalid"
        id="username"
        minLength={1}
        name="username"
        onChange={[Function]}
        type="text"
      />
    </ValidationInput>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.bitbucketcloud"
      id="personal_access_token_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_pat.bitbucketcloud"
      required={true}
    >
      <input
        autoFocus={false}
        className="input-super-large is-invalid"
        id="personal_access_token"
        minLength={1}
        onChange={[Function]}
        type="text"
        value=""
      />
    </ValidationInput>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.bitbucketcloud"
      id="personal_access_token_submit"
      isInvalid={true}
      isValid={false}
      label={null}
    >
      <SubmitButton
        disabled={true}
      >
        save
      </SubmitButton>
      <DeferredSpinner
        className="spacer-left"
        loading={false}
      />
    </ValidationInput>
  </form>
  <Alert
    className="big-spacer-left width-50"
    display="block"
    variant="info"
  >
    <h3>
      onboarding.create_project.pat_help.instructions_username.bitbucketcloud.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions_username.bitbucketcloud
    </p>
    <div
      className="text-middle big-spacer-bottom"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/bitbucket.svg"
      />
      <a
        href="https://bitbucket.org/account/settings/"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.instructions_username.bitbucketcloud.link
      </a>
    </div>
    <h3>
      onboarding.create_project.pat_help.bitbucketcloud.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      <FormattedMessage
        defaultMessage="onboarding.create_project.pat_help.bitbucketcloud.instructions"
        id="onboarding.create_project.pat_help.instructions"
        values={
          Object {
            "alm": "onboarding.alm.bitbucketcloud",
          }
        }
      />
    </p>
    <div
      className="text-middle"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/bitbucket.svg"
      />
      <a
        href="https://bitbucket.org/account/settings/app-passwords/new"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.bitbucketcloud.link
      </a>
    </div>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions2.bitbucketcloud
    </p>
    <ul>
      <li>
        <FormattedMessage
          defaultMessage="onboarding.create_project.pat_help.bbs_permission_repos"
          id="onboarding.create_project.pat_help.bbs_permission_repos"
          values={
            Object {
              "perm": <strong>
                onboarding.create_project.pat_help.read_permission
              </strong>,
            }
          }
        />
      </li>
    </ul>
  </Alert>
</div>
`;

exports[`should render correctly: gitlab 1`] = `
<div
  className="display-flex-start"
>
  <form
    className="width-50"
    onSubmit={[Function]}
  >
    <h2
      className="big"
    >
      onboarding.create_project.pat_form.title.gitlab
    </h2>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_form.help.gitlab
    </p>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.gitlab"
      id="personal_access_token_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_pat"
      required={true}
    >
      <input
        autoFocus={true}
        className="input-super-large is-invalid"
        id="personal_access_token"
        minLength={1}
        onChange={[Function]}
        type="text"
        value=""
      />
    </ValidationInput>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.gitlab"
      id="personal_access_token_submit"
      isInvalid={true}
      isValid={false}
      label={null}
    >
      <SubmitButton
        disabled={true}
      >
        save
      </SubmitButton>
      <DeferredSpinner
        className="spacer-left"
        loading={false}
      />
    </ValidationInput>
  </form>
  <Alert
    className="big-spacer-left width-50"
    display="block"
    variant="info"
  >
    <h3>
      onboarding.create_project.pat_help.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      <FormattedMessage
        defaultMessage="onboarding.create_project.pat_help.instructions"
        id="onboarding.create_project.pat_help.instructions"
        values={
          Object {
            "alm": "onboarding.alm.gitlab",
          }
        }
      />
    </p>
    <div
      className="text-middle"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/gitlab.svg"
      />
      <a
        href="https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.link
      </a>
    </div>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions2.gitlab
    </p>
    <ul>
      <li
        className="spacer-bottom"
      >
        <strong>
          onboarding.create_project.pat_help.gitlab.read_api_permission
        </strong>
      </li>
    </ul>
  </Alert>
</div>
`;

exports[`should render correctly: gitlab with non-standard api path 1`] = `
<div
  className="display-flex-start"
>
  <form
    className="width-50"
    onSubmit={[Function]}
  >
    <h2
      className="big"
    >
      onboarding.create_project.pat_form.title.gitlab
    </h2>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_form.help.gitlab
    </p>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.gitlab"
      id="personal_access_token_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_pat"
      required={true}
    >
      <input
        autoFocus={true}
        className="input-super-large is-invalid"
        id="personal_access_token"
        minLength={1}
        onChange={[Function]}
        type="text"
        value=""
      />
    </ValidationInput>
    <ValidationInput
      error="onboarding.create_project.pat_incorrect.gitlab"
      id="personal_access_token_submit"
      isInvalid={true}
      isValid={false}
      label={null}
    >
      <SubmitButton
        disabled={true}
      >
        save
      </SubmitButton>
      <DeferredSpinner
        className="spacer-left"
        loading={false}
      />
    </ValidationInput>
  </form>
  <Alert
    className="big-spacer-left width-50"
    display="block"
    variant="info"
  >
    <h3>
      onboarding.create_project.pat_help.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      <FormattedMessage
        defaultMessage="onboarding.create_project.pat_help.instructions"
        id="onboarding.create_project.pat_help.instructions"
        values={
          Object {
            "alm": "onboarding.alm.gitlab",
          }
        }
      />
    </p>
    <div
      className="text-middle"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/gitlab.svg"
      />
      <a
        href="https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.link
      </a>
    </div>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions2.gitlab
    </p>
    <ul>
      <li
        className="spacer-bottom"
      >
        <strong>
          onboarding.create_project.pat_help.gitlab.read_api_permission
        </strong>
      </li>
    </ul>
  </Alert>
</div>
`;

exports[`should render correctly: no token needed 1`] = `
<DeferredSpinner
  className="spacer-left"
  loading={true}
/>
`;

exports[`should show error when issue: issue submitting token 1`] = `
<div
  className="display-flex-start"
>
  <form
    className="width-50"
    onSubmit={[Function]}
  >
    <h2
      className="big"
    >
      onboarding.create_project.pat_form.title.bitbucketcloud
    </h2>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_form.help.bitbucketcloud
    </p>
    <ValidationInput
      id="enter_username_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_username"
      required={true}
    >
      <input
        autoFocus={true}
        className="input-super-large is-invalid"
        id="username"
        minLength={1}
        name="username"
        onChange={[Function]}
        type="text"
        value="username"
      />
    </ValidationInput>
    <ValidationInput
      error="default_error_message"
      id="personal_access_token_validation"
      isInvalid={false}
      isValid={false}
      label="onboarding.create_project.enter_pat.bitbucketcloud"
      required={true}
    >
      <input
        autoFocus={false}
        className="input-super-large is-invalid"
        id="personal_access_token"
        minLength={1}
        onChange={[Function]}
        type="text"
        value="token"
      />
    </ValidationInput>
    <ValidationInput
      error="default_error_message"
      id="personal_access_token_submit"
      isInvalid={true}
      isValid={false}
      label={null}
    >
      <SubmitButton
        disabled={true}
      >
        save
      </SubmitButton>
      <DeferredSpinner
        className="spacer-left"
        loading={false}
      />
    </ValidationInput>
  </form>
  <Alert
    className="big-spacer-left width-50"
    display="block"
    variant="info"
  >
    <h3>
      onboarding.create_project.pat_help.instructions_username.bitbucketcloud.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions_username.bitbucketcloud
    </p>
    <div
      className="text-middle big-spacer-bottom"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/bitbucket.svg"
      />
      <a
        href="https://bitbucket.org/account/settings/"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.instructions_username.bitbucketcloud.link
      </a>
    </div>
    <h3>
      onboarding.create_project.pat_help.bitbucketcloud.title
    </h3>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      <FormattedMessage
        defaultMessage="onboarding.create_project.pat_help.bitbucketcloud.instructions"
        id="onboarding.create_project.pat_help.instructions"
        values={
          Object {
            "alm": "onboarding.alm.bitbucketcloud",
          }
        }
      />
    </p>
    <div
      className="text-middle"
    >
      <img
        alt=""
        className="spacer-right"
        height="16"
        src="/images/alm/bitbucket.svg"
      />
      <a
        href="https://bitbucket.org/account/settings/app-passwords/new"
        rel="noopener noreferrer"
        target="_blank"
      >
        onboarding.create_project.pat_help.bitbucketcloud.link
      </a>
    </div>
    <p
      className="big-spacer-top big-spacer-bottom"
    >
      onboarding.create_project.pat_help.instructions2.bitbucketcloud
    </p>
    <ul>
      <li>
        <FormattedMessage
          defaultMessage="onboarding.create_project.pat_help.bbs_permission_repos"
          id="onboarding.create_project.pat_help.bbs_permission_repos"
          values={
            Object {
              "perm": <strong>
                onboarding.create_project.pat_help.read_permission
              </strong>,
            }
          }
        />
      </li>
    </ul>
  </Alert>
</div>
`;
